plugins/module_utils/foreman_helper.py

```diff
@@ -471,6 +483,77 @@ def find_puppetclass(self, name, environment=None, params=None, failsafe=False,
+            return self.foreman_params[key]
+
+        resource_type = entity_spec['resource_type']
+        failsafe = entity_spec.get('failsafe', False)
+        thin = entity_spec.get('thin', True)
+        params = {}
+        try:
+            if 'scope' in entity_spec:
+                for scope in entity_spec['scope']:
+                    params.update(self.scope_for(scope))
+        except TypeError:
+            if failsafe:
+                if entity_spec.get('type') == 'entity':
+                    result = None
+                else:
+                    result = [None for value in self.foreman_params[key]]
+            else:
+                self.fail_json(msg="Failed to lookup scope {0} while searching for {1}.".format(entity_spec['scope'], resource_type))
+        else:
+            # No exception happend => scope is in place
+            if resource_type == 'operatingsystems':
+                if entity_spec.get('type') == 'entity':
+                    result = self.find_operatingsystem(self.foreman_params[key], params=params, failsafe=failsafe, thin=thin)
+                else:
+                    result = [self.find_operatingsystem(value, params=params, failsafe=failsafe, thin=thin) for value in self.foreman_params[key]]
+            elif resource_type == 'puppetclasses':
+                if entity_spec.get('type') == 'entity':
+                    result = self.find_puppetclass(self.foreman_params[key], params=params, failsafe=failsafe, thin=thin)
+                else:
+                    result = [self.find_puppetclass(value, params=params, failsafe=failsafe, thin=thin) for value in self.foreman_params[key]]
+            else:
+                if entity_spec.get('type') == 'entity':
+                    result = self.find_resource_by(
+                        resource=resource_type,
+                        value=self.foreman_params[key],
+                        search_field=entity_spec.get('search_by', ENTITY_KEYS.get(resource_type, 'name')),
+                        search_operator=entity_spec.get('search_operator', '='),
+                        failsafe=failsafe, thin=thin, params=params,
+                    )
+                else:
+                    result = [self.find_resource_by(
+                        resource=resource_type,
+                        value=value,
+                        search_field=entity_spec.get('search_by', ENTITY_KEYS.get(resource_type, 'name')),
+                        search_operator=entity_spec.get('search_operator', '='),
+                        failsafe=failsafe, thin=thin, params=params,
+                    ) for value in self.foreman_params[key]]
+        self.set_entity(key, result)
+        return result
+
+    def auto_lookup_entities(self):
+        return [
+            self.lookup_entity(key)
+            for key, entity_spec in self.foreman_spec.items()
+            if entity_spec.get('resolve', True) and entity_spec.get('type') in {'entity', 'entity_list'}
```

Comment:
Out of curiosity, why are you using a set (`in {a, b}`) and not a list (`in [a, b]`) or a tuple (`in (a, b)`). Is set access quicker?

Reply:
Superstition. In theory lookup in a set should be quicker. Building it might not be, and for two elements, it may not be worth discussing.